### PR TITLE
backend: Refactor Retriever/Function tools into single Tool concept

### PR DESCRIPTION
**Pull request by tianjing-li**
- Removes the notion of Retriever or Function tools, they are now all considered tools
- Now all tools should inherit from `BaseTool`
- Updates all existing config and imports
- Updates unit tests

### CODE REVIEW

File: src/backend/tools/lang_chain.py [hunk 1]
```diff
@@ -31,13 +31,14 @@ def __init__(self, chunk_size: int = 300, chunk_overlap: int = 0):
     def is_available(cls) -> bool:
         return True
 
-    def retrieve_documents(self, query: str, **kwargs: Any) -> List[Dict[str, Any]]:
+    def call(self, parameters: str, **kwargs: Any) -> List[Dict[str, Any]]:
         wiki_retriever = WikipediaRetriever()
-        docs = wiki_retriever.get_relevant_documents(query)
+        docs = wiki_retriever.get_relevant_documents(parameters)
```
Comment (BeatrixCohere) [inline src/backend/tools/lang_chain.py:36]: We need to extract "query" from parameters

File: src/backend/tools/lang_chain.py [hunk 2]
```diff
@@ -62,13 +63,16 @@ def __init__(self, filepath: str):
...
+
         # Load text files and split into chunks
         loader = PyPDFLoader(self.filepath)
         text_splitter = CharacterTextSplitter(chunk_size=300, chunk_overlap=0)
         pages = loader.load_and_split(text_splitter)
+
         # Create a vector store from the documents
         db = Chroma.from_documents(documents=pages, embedding=cohere_embeddings)
-        input_docs = db.as_retriever().get_relevant_documents(query)
+        input_docs = db.as_retriever().get_relevant_documents(parameters)
```
Comment (BeatrixCohere) [inline src/backend/tools/lang_chain.py:76]: Same here, we need to extract query from parameters

File: src/backend/tools/tavily.py [hunk 3]
```diff
@@ -17,8 +17,8 @@ def __init__(self):
     def is_available(cls) -> bool:
         return cls.tavily_api_key is not None
 
-    def retrieve_documents(self, query: str, **kwargs: Any) -> List[Dict[str, Any]]:
-        content = self.client.search(query=query, search_depth="advanced")
+    def call(self, parameters: str, **kwargs: Any) -> List[Dict[str, Any]]:
+        content = self.client.search(query=parameters, search_depth="advanced")
```
Comment (BeatrixCohere) [inline src/backend/tools/tavily.py:21]: Same here for "query"

File: src/backend/config/tools.py [hunk 4]
```diff
@@ -4,13 +4,11 @@
...
-    PythonInterpreterFunctionTool,
-)
-from backend.tools.retrieval import (
+from backend.tools import (
+    Calculator,
     LangChainVectorDBRetriever,
     LangChainWikiRetriever,
+    PythonInterpreter,
     TavilyInternetSearch,
 )
```
Comment (BeatrixCohere) [inline src/backend/config/tools.py:14]: We need to add parameter definitions for the new tools from "retrievers"

File: src/backend/tools/lang_chain.py (hunk 1 above)
Comment (tianjing-li) [reply src/backend/tools/lang_chain.py:36]: Done, modified so I pass in a query param

File: docs/custom_tool_guides/tool_guide.md [hunk 5]
```diff
@@ -25,115 +25,48 @@ There are three types of tools:
...
-If you need to install a new module to run your tool, execute the following command and run `make dev` again.
+If you need to install a new library to run your tool, execute the following command and run `make dev` again.
 
 ```bash
 poetry add <MODULE> --group community
 ```
+### Implementing a Tool
 
-If you're working on a File or Data Loader, follow the steps outlined in [Implementing a Retriever](#implementing-a-retriever).
+Add the implementation inside a tool class that inherits from `BaseTool`. This class will need to implement
```
Comment (lusmoura) [inline docs/custom_tool_guides/tool_guide.md:37]: Should we mention that the UI uses the `text` field to populate the citations? I've seen people implementing tools that didn't return `text` confused with the UI

**Review by lusmoura — APPROVED** [review]
🚀 just one nit

File: docs/custom_tool_guides/tool_guide.md (hunk 5 above)
Comment (tianjing-li) [reply docs/custom_tool_guides/tool_guide.md:37]: Good point

Comment (tianjing-li) [reply docs/custom_tool_guides/tool_guide.md]: @lusmoura done

**Review by lusmoura — APPROVED** [review]

**Pull request merged**